core/embed/extmod/modtrezorio/modtrezorio-flash.h

```diff
@@ -118,3 +123,302 @@ STATIC const mp_obj_type_t mod_trezorio_FlashOTP_type = {
+    if (i % PROGRESS_CHUNK_SIZE == 0) {
+      ui_progress(ui_wait_callback, i * FLASH_READ_CHUNK_SIZE);
+    }
+  }
+
+  ui_progress(ui_wait_callback, length);
+
+  vstr_t vstr = {0};
+  vstr_init_len(&vstr, BLAKE2S_DIGEST_LENGTH);
+  if (blake2s_Final(&ctx, vstr.buf, vstr.len) != 0) {
+    vstr_clear(&vstr);
+    mp_raise_msg(&mp_type_RuntimeError, "Failed to finalize hash.");
+  }
+
+  return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorio_FlashArea_hash_obj, 3,
+                                           5, mod_trezorio_FlashArea_hash);
+
+/// if __debug__:
+///     def read(self, offset: int, data: bytearray) -> None:
+///         """
+///         Reads data from flash area. Will read exact length of data
+///         bytearray. Offset and length of data must be aligned to 1024 bytes.
+///         """
+#if DEBUG
+STATIC mp_obj_t mod_trezorio_FlashArea_read(mp_obj_t obj_self,
+                                            mp_obj_t obj_offset,
+                                            mp_obj_t obj_data) {
+  mp_obj_FlashArea_t *self = (mp_obj_FlashArea_t *)MP_OBJ_TO_PTR(obj_self);
+  uint32_t offset = trezor_obj_get_uint(obj_offset);
+  mp_buffer_info_t data = {0};
+  mp_get_buffer_raise(obj_data, &data, MP_BUFFER_WRITE);
+
+  if (offset % FLASH_READ_CHUNK_SIZE != 0 ||
+      data.len % FLASH_READ_CHUNK_SIZE != 0) {
+    mp_raise_ValueError("Offset and length must be aligned to 1024 bytes.");
+  }
+
+  uint32_t area_size = flash_area_get_size(self->area);
+  if (area_size - offset < data.len) {
+    mp_raise_ValueError("Read too long.");
+  }
+  uint32_t chunks = data.len / FLASH_READ_CHUNK_SIZE;
+  for (int i = 0; i < chunks; i++) {
+    const void *data = flash_area_get_address(
+        &FIRMWARE_AREA, i * FLASH_READ_CHUNK_SIZE, FLASH_READ_CHUNK_SIZE);
+    if (data == NULL) {
+      mp_raise_msg(&mp_type_RuntimeError, "Failed to read flash.");
+    }
+    memcpy(data.buf + i * FLASH_READ_CHUNK_SIZE, data, FLASH_READ_CHUNK_SIZE);
+  }
+  return mp_const_none;
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(mod_trezorio_FlashArea_read_obj,
+                                 mod_trezorio_FlashArea_read);
+
+/// if __debug__:
+///     def write(self, offset: int, data: bytes) -> None:
+///         """
+///         Writes data to flash area.
+///         Offset and written data size must be a multiple of quadword (16
+///         bytes).
+///         """
+STATIC mp_obj_t mod_trezorio_FlashArea_write(mp_obj_t obj_self,
+                                             mp_obj_t obj_offset,
+                                             mp_obj_t obj_data) {
+  mp_obj_FlashArea_t *self = (mp_obj_FlashArea_t *)MP_OBJ_TO_PTR(obj_self);
+  uint32_t offset = trezor_obj_get_uint(obj_offset);
+  mp_buffer_info_t data = {0};
+  mp_get_buffer_raise(obj_data, &data, MP_BUFFER_READ);
+
+  if (data.len % FLASH_BLOCK_SIZE != 0) {
+    mp_raise_ValueError("Write size must be a multiple of 16.");
```

Comment:
16 is wrong on F4

Reply:
c8db36e3fa40721edad402982a1c3f262d2b30b6